src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs

```diff
@@ -394,54 +393,94 @@ public void HandleLogMessage(TestMessageLevel level, string message)
         /// <param name="rawMessage">Raw message.</param>
         public void HandleRawMessage(string rawMessage)
         {
-            if (this.requestData.IsTelemetryOptedIn)
+            // Note: Deserialize rawMessage only if required.
+
+            var message = this.LoggerManager.LoggersInitialized || this.requestData.IsTelemetryOptedIn ?
+                    this.dataSerializer.DeserializeMessage(rawMessage) : null;
+
+            if (string.Equals(message?.MessageType, MessageType.DiscoveryComplete))
             {
-                var message = this.dataSerializer.DeserializeMessage(rawMessage);
+                var discoveryCompletePayload = this.LoggerManager.LoggersInitialized || this.requestData.IsTelemetryOptedIn ?
```

Comment:
>this.LoggerManager.LoggersInitialized || this.requestData.IsTel [](start = 47, length = 63)

are these checks needed again? You have already checked above that either Loggers or telemetry is enabled

Reply:
removed